src/MizarWidgets/include/MizarWidgets/SamplingWithFrameTrackInputWidget.h

```diff
@@ -5,30 +5,92 @@
 
+#include <absl/container/flat_hash_map.h>
+#include <absl/container/flat_hash_set.h>
+#include <stdint.h>
+
+#include <QLabel>
+#include <QListWidget>
+#include <QObject>
 #include <QWidget>
 #include <memory>
 #include <string_view>
+#include <vector>
 
 #include "MizarData/MizarPairedData.h"
+#include "MizarData/SamplingWithFrameTrackComparisonReport.h"
 
 namespace Ui {
 class SamplingWithFrameTrackInputWidget;
 }
 
 namespace orbit_mizar_widgets {
 
-class SamplingWithFrameTrackInputWidget : public QWidget {
+class SamplingWithFrameTrackInputWidgetBase : public QWidget {
+  Q_OBJECT
  public:
-  explicit SamplingWithFrameTrackInputWidget(QWidget* parent);
-  ~SamplingWithFrameTrackInputWidget() override;
+  ~SamplingWithFrameTrackInputWidgetBase() override;
+
+  [[nodiscard]] orbit_mizar_data::HalfOfSamplingWithFrameTrackReportConfig MakeConfig() const;
 
-  template <typename PairedData = orbit_mizar_data::MizarPairedData>
-  void Init(const PairedData* data, const QString& name);
+ public slots:
+  void OnThreadSelectionChanged();
 
  private:
   std::unique_ptr<Ui::SamplingWithFrameTrackInputWidget> ui_;
+
+ protected:
+  explicit SamplingWithFrameTrackInputWidgetBase(QWidget* parent);
+
+  [[nodiscard]] QLabel* GetTitle() const;
+  [[nodiscard]] QListWidget* GetThreadList() const;
+
+  absl::node_hash_map<std::unique_ptr<QListWidgetItem>, uint32_t> tid_list_widget_items_to_tids_;
+
+ private:
+  absl::flat_hash_set<uint32_t> selected_tids_;
 };
 
+template <typename PairedData>
+class SamplingWithFrameTrackInputWidgetTmpl : public SamplingWithFrameTrackInputWidgetBase {
+ public:
+  SamplingWithFrameTrackInputWidgetTmpl() = delete;
+  explicit SamplingWithFrameTrackInputWidgetTmpl(QWidget* parent)
+      : SamplingWithFrameTrackInputWidgetBase(parent) {}
+  ~SamplingWithFrameTrackInputWidgetTmpl() override = default;
+
+  void Init(const PairedData& data, const QString& name) {
+    InitTitle(name);
+    InitThreadList(data);
+  }
+
+ private:
+  void InitTitle(const QString& name) { GetTitle()->setText(name); }
+
+  void InitThreadList(const PairedData& data) {
+    QListWidget* list = GetThreadList();
+    list->setSelectionMode(QAbstractItemView::ExtendedSelection);
+
+    absl::flat_hash_map<uint32_t, std::string> tid_to_name = data.TidToNames();
+    absl::flat_hash_map<uint32_t, uint64_t> counts = data.TidToCallstackSampleCounts();
+
+    std::vector<std::pair<uint32_t, uint64_t>> counts_sorted(std::begin(counts), std::end(counts));
+
+    std::sort(std::begin(counts_sorted), std::end(counts_sorted),
+              [](const auto& a, const auto& b) { return a.second > b.second; });
+
+    for (const auto& [tid, unused_count] : counts_sorted) {
+      auto item = std::make_unique<QListWidgetItem>(
+          QString::fromStdString(absl::StrFormat("[%u] %s", tid, tid_to_name.at(tid))));
+      list->addItem(item.get());
+      tid_list_widget_items_to_tids_.try_emplace(std::move(item), tid);
```

Comment:
I am unsure if this is safe. I think with `addItem`, the `QListWidget` takes ownership of the `QListWidgetItem`. Maybe this can then lead to a situation where the item is destructed/deleted twice? I'd also like to understand this more. @beckerhe, any experience here?

Reply:
> Maybe this can then lead to a situation where the item is destructed/deleted twice? 

Doesn't seem so. There was an issue with QListWidgetItem emitting signals upon destruction which lead to the need to carefully choose field destruction order. But yeah, probably, there's another way.

Reply:
The proper way would be to define a model.

But it also be possible to define your own role like that:

```cpp
enum CustomRoles {
  kTidRole = Qt::UserRole + 1
};
```

and then store the value as custom data in the item:

```cpp
listItem->setData(kTidRole, tid);
```

and retrieve it if needed:

```cpp
auto tid = listItem->data(kTidRole).value<uint32_t>();
```

Reply:
In the other comment I suggested subclassing `QWidgetListItem`, because I missed this `setData` method when reading the Qt docs. I like Hennings suggestion and am in favor of implementing it this way.

Reply:
Also.. I still would like to understand what exactly is happening here with destruction.